### PR TITLE
Sliding sync: Add connection tracking to the `account_data` extension

### PR DESCRIPTION
**Pull request by erikjohnston**
This is basically exactly the same logic as for receipts. Essentially we just need to track which room account data we have and haven't sent down to clients, and use that when we pull stuff out.

I think this just needs a couple of extra tests written

### CODE REVIEW

File: synapse/handlers/sliding_sync/extensions.py [hunk 1]
```diff
@@ -456,58 +485,109 @@ async def get_account_data_extension_response(
...
+                            room_id=room_id,
+                            from_stream_id=previous_token,
+                            to_stream_id=to_token.account_data_key,
+                        )
+                        if changed:
+                            # XXX: Ideally, this should take into account the `to_token`
+                            # and return the set of tags at that time but we don't track
+                            # changes to tags so we just have to return all tags for the
+                            # room.
+                            immutable_tag_map = await self.store.get_tags_for_room(
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/extensions.py:528]: Cache for `get_tags_for_room(...)` is being added in https://github.com/element-hq/synapse/pull/17730

File: synapse/handlers/sliding_sync/extensions.py [hunk 2]
```diff
@@ -456,58 +485,109 @@ async def get_account_data_extension_response(
...
+                                user_id, room_id
+                            )
+                            if immutable_tag_map:
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync/extensions.py:531]: hmm, I think you're right. It does make sense to show an empty set of tags in an incremental sync if they changed

File: synapse/handlers/sliding_sync/extensions.py [hunk 3]
```diff
@@ -456,58 +485,109 @@ async def get_account_data_extension_response(
...
+                        # changes to tags so we just have to return all tags for the
+                        # room.
+                        immutable_tag_map = await self.store.get_tags_for_room(
+                            user_id, room_id
+                        )
+
+                        account_data_by_room_map[room_id] = ChainMap(
+                            {AccountDataTypes.TAG: {"tags": immutable_tag_map}}
+                            if immutable_tag_map
+                            else {},
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/extensions.py:558]: Do we want to always return tags in an intial sync?

The problem is we can't tell no tag account data from empty set of tags.

Comment (erikjohnston) [reply synapse/handlers/sliding_sync/extensions.py:558]: We probably don't? Though I don't think it really matters

File: tests/rest/client/sliding_sync/test_extension_account_data.py [hunk 4]
```diff
@@ -480,6 +482,337 @@ def test_room_account_data_incremental_sync(self) -> None:
             {"tags": {"m.favourite": {}, "m.server_notice": {}}},
         )
 
+    def test_room_account_data_incremental_sync_out_of_range_never(self) -> None:
+        """Tests that we don't return account data for rooms that fall out of
+        range, but then do send all account data once they're back in range.
```
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply tests/rest/client/sliding_sync/test_extension_account_data.py:487]: Updated

> Tests that we don't return account data for rooms that are out of
> range, but then do send all account data once they're in range.

File: synapse/handlers/sliding_sync/extensions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync/extensions.py]: Updated, check if the tests match your expectations

File: synapse/handlers/sliding_sync/extensions.py (hunk 3 above)
Comment (MadLittleMods) [reply synapse/handlers/sliding_sync/extensions.py:557]: I think I got it right. Only show the empty tag set if we previously told the client about tags. Check if the tests match your expectations

**Review by erikjohnston — COMMENTED** [review]
LGTM, but I can't approve as i opened the PR...

**Comment by MadLittleMods** [comment]
Feels like I shouldn't self-approve my own work :thinking:

**Review by MadLittleMods — APPROVED** [review]
Approving on behalf of @erikjohnston

**Pull request merged**